### PR TITLE
Add `pauli_sentence()` function which takes native PennyLane Ops to `PauliSentence` instance (PauliUtility #2)

### PR DESCRIPTION
**Pull request by Jaybsoni**
**Context:**
`pauli_sentence` is a utility function which uses single dispatch to take existing PennyLane Operators and returns an instance of `PauliSentence` class.

**Description of the Change:**
- Update `is_pauli_word` to support instances of the hamiltonian class
- Introduce `pauli_sentence` function + tests 
- Fixed (independent) bug where calling for wires of an empty PauliWord caused errors in the to_mat method for PauliSentence

**Benefits:**
- Can now interface between native PennyLane ops and the new lightweight classes

### CODE REVIEW

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -46,6 +46,24 @@
 
 * New basis sets, `6-311g` and `CC-PVDZ`, are added to the qchem basis set repo.
   [#3279](https://github.com/PennyLaneAI/pennylane/pull/3279)
+  
+* New `pauli_sentence()` function which takes native `Operator` or `Hamiltonian`
+  instances representing a linear combination of Pauli words and returns 
+  the equivalent `PauliSentence`.
+  [(#3252)](https://github.com/PennyLaneAI/pennylane/pull/3389)
```
Comment (Jaybsoni) [inline doc/releases/changelog-dev.md:53]:
```suggestion
  [(#3389)](https://github.com/PennyLaneAI/pennylane/pull/3389)
```

File: pennylane/pauli/utils.py [hunk 2]
```diff
@@ -66,8 +66,7 @@ def is_pauli_word(observable):
 
 
     Args:
-        observable (Observable): an observable, either a :class:`~.Tensor` instance or
-            single-qubit observable.
+        observable (Union[~.Observable]): the operator to be examined
```
Comment (AlbertMitjans) [inline pennylane/pauli/utils.py:69]: No need of `Union` if we only have one type

File: pennylane/pauli/utils.py [hunk 3]
```diff
@@ -81,14 +80,22 @@ def is_pauli_word(observable):
...
     if isinstance(observable, Tensor):
         return set(observable.name).issubset(pauli_word_names)
 
+    if isinstance(observable, Hamiltonian):
+        terms_pauli_word = []
+        for _, ob in zip(*observable.terms()):
+            if isinstance(ob, Tensor):
+                terms_pauli_word.append(set(ob.name).issubset(pauli_word_names))
+            else:
+                terms_pauli_word.append(ob.name in pauli_word_names)
```
Comment (AlbertMitjans) [inline pennylane/pauli/utils.py:96]: I believe can use recursion here:
```suggestion
            terms_pauli_word.append(is_pauli_word(ob))
```

File: pennylane/pauli/utils.py [hunk 4]
```diff
@@ -1145,7 +1152,9 @@ def diagonalize_pauli_word(pauli_word):
     return diag_term
 
 
-def diagonalize_qwc_pauli_words(qwc_grouping):  # pylint: disable=too-many-branches
+def diagonalize_qwc_pauli_words(
+    qwc_grouping,
+):  # pylint: disable=too-many-branches, isinstance-second-argument-not-valid-type
```
Comment (AlbertMitjans) [inline pennylane/pauli/utils.py:1157]: Why is this needed? Pylint doesn't raise any error for me.

Comment (AlbertMitjans) [reply pennylane/pauli/utils.py:1157]: And all types used in the added isinstance checks should be valid...

**Review by AlbertMitjans — COMMENTED** [review]
Nice work! 🚀

File: pennylane/pauli/conversion.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    if not is_pauli_word(op):
+        raise ValueError(f"Op must be a linear combination of Pauli operators only, got: {op}")
+
+    pw = {}
+    for factor in op.obs:
+        wire, pauli = (factor.wires[0], op_to_str_map[factor.__class__])
+        pw[wire] = pauli
+
+    return PauliSentence({PauliWord(pw): 1.0})
+
```
Comment (lillian542) [inline pennylane/pauli/conversion.py:72]: The conversion of a `Tensor` to a `PauliSentence` here produces an odd result if it's being used in Tensors with overlapping wires, ie.

```
op = qml.PauliX(0) @ qml.PauliY(0)
pauli_sentence(op)
>>> 1.0 * Y(0)
```
Same thing for a `Hamiltonian` defined based on a `Tensor` of multiple Pauli operators acting on the same wire.

If we fix #3334 so that `is_pauli_word` confirms that an operator is truly a Pauli word, this will raise a ValueError instead (and we should reword the error message a little). That would be fine, but it would be inconsistent with what happens if you try to do the same thing with `Prod`, which handles it nicely:

```
op = prod(qml.PauliX(0) @ qml.PauliY(0))
pauli_sentence(op)
>>> 1j * Z(0)
```

Would it be weird to convert it into a `Prod` instead of processing it as a `Tensor`? `Prod` behaves better and the output is all `PauliWord` and `PauliSentence` objects either way.

File: pennylane/pauli/utils.py (hunk 2 above)
Comment (Jaybsoni) [reply pennylane/pauli/utils.py:69]: Correct, I think this is a relic from another PR that I split to create this one 😄

**Comment by Jaybsoni** [comment]
[sc-28974]

File: pennylane/pauli/utils.py (hunk 4 above)
Comment (Jaybsoni) [reply pennylane/pauli/utils.py:1157]: Yes, but I think the issue is the tuple. For some reason pylint thinks we are comparing the op to the class `tuple` even though the function accepts an iterable in the second argument.

It is described [here](https://github.com/PyCQA/pylint/issues/3507), I think we need to upgrade pylint.

File: pennylane/pauli/utils.py (hunk 3 above)
Comment (Jaybsoni) [reply pennylane/pauli/utils.py:96]: Nice catch!

File: pennylane/pauli/conversion.py (hunk 5 above)
Comment (Jaybsoni) [reply pennylane/pauli/conversion.py:72]: We definitely need to fix `is_pauli_word` but I think we should do that in a separate PR. For now I update the logic to take this case into account when constructing from a tensor and add test cases for this.

File: doc/releases/changelog-dev.md [hunk 6]
```diff
@@ -46,6 +46,24 @@
...
+  >>> h = qml.Hamiltonian([1.23], [qml.PauliX(wires=0) @ qml.PauliZ(wires=1)])
+  >>> print(h)
+    (1.23) [X0 Z1]
+  >>> qml.pauli.pauli_sentence(op)
+  1.23 * Z(1) @ X(0)
+  >>> qml.pauli.pauli_sentence(h)
+  1.23 * X(0) @ Z(1)
+  ```
 
 <h3>Improvements</h3>
```
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply doc/releases/changelog-dev.md:159]: Added 👍🏼

File: pennylane/pauli/conversion.py (hunk 5 above)
Comment (Jaybsoni) [reply pennylane/pauli/conversion.py:68]: Added 👍🏼

**Review by lillian542 — APPROVED** [review]
Looks good to me!

File: pennylane/pauli/utils.py (hunk 4 above)
Comment (AlbertMitjans) [reply pennylane/pauli/utils.py:1154]: Where was the error raised, in your local `pre-commit` or in the github action?

If I remember correctly, I used several times a `tuple` in the `isinstance` check and I never got a complain from pylint...

**Review by AlbertMitjans — APPROVED** [review]
Good job!

File: pennylane/pauli/utils.py (hunk 4 above)
Comment (Jaybsoni) [reply pennylane/pauli/utils.py:1154]: In GitHub actions, I can undo the change and we can double check.

File: pennylane/pauli/pauli_arithmetic.py [hunk 7]
```diff
@@ -13,6 +13,7 @@
 # limitations under the License.
 """The Pauli arithmetic abstract reduced representation classes"""
 from copy import copy
+from typing import Sequence
```
Comment (Jaybsoni) [inline pennylane/pauli/pauli_arithmetic.py:16]:
```suggestion
```
